### PR TITLE
fix(responses): share one progressive freeform decoder with routed restoration

### PR DESCRIPTION
**Pull request by lidge-jun**
## Summary

- On the routed Responses path, a custom-tool preview could publish bytes the authoritative final input does not contain. The progressive decoder emitted the string inside the `input` wrapper immediately, while completion strips a complete outer Markdown fence for `exec` and `apply_patch`. A delta consumer therefore received fence bytes that completion then removed, and no SSE event can take a published prefix back. The bridge fix in #5070 did not reach this module.
- The decoder now lives in `src/responses/progressive-freeform-input.ts`, and both paths call it. Existing bridge wrapper/fence behavior and its separate patch-envelope hold are preserved. The shared decoder additionally suppresses a wrapper containing literal JSON control characters, which completion cannot parse.
- Routed restoration keeps the holds it already had — an unrecognized JSON object, and a code-mode patch envelope — and keeps owning its retained bytes. The shared decoder is pure: it charges, releases and retains nothing, so there is no second `TranslatorBudget` owner and the existing charge/release lifecycle is untouched.
- Ordinary raw input stays progressive. A prefix that can still become a fence is held for the whole call, because a closing fence can still be followed by text that withdraws it.
- Closes #5113.

Two limits are documented rather than claimed away. A duplicate `input` key resolves to the last one at completion, and a wrapper that becomes invalid only after a valid prefix was emitted cannot retract that prefix. Both are inherent to progressive delivery while canonical input stays progressive; the completed item remains authoritative, and the new test pins both.

Scope map:

| Path | Change |
| --- | --- |
| `src/responses/progressive-freeform-input.ts` | New. Pure shared decoder with literal-control rejection; existing wrapper/fence policy preserved. |
| `src/bridge/sse.ts` | Calls the shared decoder; local implementation removed, behavior unchanged. |
| `src/server/responses-custom-tool-repair.ts` | Calls the shared decoder with the tool grammar it owns; its own holds preserved. |
| `tests/responses/responses-custom-tool-stream-consistency.test.ts` | New sibling file, registered in both layout inventories. |
| `structure/transports/responses.md`, `byte-accounting.md`, `inventory.md` | Record the shared decoder, its purity, and the documented limits. |

The existing `tests/responses/responses-custom-tool-repair.test.ts` is exactly at its 2,143-line ratchet cap and is unchanged at 2,143 lines, which is why the coverage is a sibling file.

## Verification

- Local checks: **NOT RUN.** This lane is prohibited from running local suites, focused tests, typecheck, builds, installs or any live `ocx` command, so no local result is claimed. Exact-head hosted CI on this PR is the executable evidence.
- Static review performed instead: compared the shared module against the former bridge decoder, preserving its call contract while reviewing the intentional literal-control rejection; confirmed the new module imports only `freeformFallbackKeys` and `unwrapFreeformToolInput` and references no budget API; confirmed the routed module still holds an unrecognized JSON object, so the existing assertion covering that case is unaffected.
- New regressions cover a canonical wrapped fenced `exec` body at every split boundary, ordinary raw input staying progressive, fallback keys, JSON whitespace, escapes, a split surrogate pair, stable call and item identity, and the two documented limits.
- `git diff --check` is clean. New test file registered in `scripts/test-layout/layout.json` and `tests/fixtures/test-layout-expected.json`; no file-size cap was raised.

## Checklist

- [x] Scope stays focused and avoids unrelated cleanup.
- [x] Docs or release notes were updated when needed.
- [x] Security-sensitive changes were reviewed for secrets, auth, and unsafe defaults.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved progressive previews for custom tool inputs, including wrapped JSON, escaped characters, and split Unicode sequences.
  * Prevented incomplete `exec` and `apply_patch` code fences from appearing prematurely.
  * Improved consistency between streamed previews and completed tool-call input.

* **Documentation**
  * Added documentation covering preview streaming behavior and buffer accounting.

* **Tests**
  * Added coverage for custom-tool streaming across split boundaries, fallback fields, escapes, and edge cases.

### CODE REVIEW

**Comment by lidge-jun** [comment]
## 리뷰 · 우선순위 77 / 80

이 PR은 라우티드 Responses 경로에서 커스텀 툴 미리보기(preview)가 최종 입력과 어긋나던 구멍을 막습니다. 예전에는 `input` 래퍼 안의 문자열을 바로 스트리밍했는데, 완료 시점에는 `exec`/`apply_patch`의 바깥 마크다운 펜스를 떼어 버립니다. 그래서 델타 소비자는 펜스 바이트를 먼저 받고, 완료 이벤트는 그 바이트가 없는 `text(1)` 같은 값을 줍니다. SSE에는 “이미 보낸 접두사를 취소”하는 이벤트가 없습니다. 브리지 쪽 같은 문제는 #5070에서 고쳤지만, `responses-custom-tool-repair.ts`의 라우티드 복원에는 그 고침이 닿지 않았습니다. 이슈 #5113을 닫습니다.

고침 방식은 “같은 디코더를 두 경로가 공유”입니다. 브리지에 있던 점진 디코더를 `src/responses/progressive-freeform-input.ts`로 그대로 빼내고, `src/bridge/sse.ts`는 같은 인자·같은 호출 자리에서 `progressiveFreeformInput`만 부르도록 바꿨습니다. 브리지 동작은 의도상 동일합니다. 라우티드 복원은 그 공유 함수를 쓰되, 예전에 갖고 있던 더 엄한 홀드(알아보지 못한 JSON 객체, 코드모드 패치 봉투)와 바이트 charge/release 소유권은 그대로 둡니다. 공유 디코더는 순수 함수라 `TranslatorBudget`을 건드리지 않습니다. 일반 raw 입력은 계속 점진적으로 나가고, 펜스가 될 수 있는 접두사는 호출 내내 붙잡습니다.

테스트는 기존 `responses-custom-tool-repair.test.ts`가 2,143줄 상한에 딱 붙어 있어서 형제 파일 `responses-custom-tool-stream-consistency.test.ts`로 넣었습니다. 펜스 래핑 exec의 모든 split 경계, raw 진행성, fallback 키, 공백 래퍼, JSON escape, 서로게이트 쌍, 그리고 문서화된 한계 두 가지(중복 `input` 키, 유효 접두사 발행 뒤 래퍼 무효화)를 고정합니다. structure/transports 문서와 layout 등록도 맞춰 두었습니다. types/config 대형 스플릿과 무관합니다.

경로 `src/responses/progressive-freeform-input.ts` - 브리지에서 추출한 순수 디코더. 예산 API 참조 없음. 설계가 맞다
경로 `src/bridge/sse.ts` - 로컬 구현 제거 후 동일 호출. 패치 봉투 홀드는 브리지에 남김
경로 `src/server/responses-custom-tool-repair.ts` (~293-320) - 공유 디코더 + `fullInput === argumentsText && trimStart().startsWith("{")` 라우티드 전용 JSON 홀드. 의도된 더 엄한 규칙이다 문서화됨
경로 `tests/responses/responses-custom-tool-stream-consistency.test.ts` - 펜스 split 전수·문서화 한계까지 고정. 회귀에 적절하다
호스티드 CI - 이 글을 쓰는 시점 대부분 pending. 작성자 말대로 로컬 스위트는 이 레인에서 돌리지 않음. merge 전에 그린을 기다려야 한다
라인 문제(차단급) - 이 diff에서 발명할 차단 버그는 없다. 중복 `input`·늦은 무효 래퍼는 문서화된 한계이며 테스트가 핀으로 박혀 있다

메인테이너의 판단이 필요한 지점
- 호스티드 CI(특히 responses 관련 test shard·structure gate)가 그린인지 확인한 뒤 merge할지
- 라우티드 경로가 `{`로 시작하는 비-래퍼 raw 바디까지 done까지 홀드하는 역사적 규칙(공유 디코더가 raw를 흘려보내므로 별도 if로 유지)을 앞으로도 유지할지
- #5070 브리지 수정과의 “한 디코더” 계약이 이후 브리지-only 핫픽스에서 다시 갈라지지 않게 리뷰 체크리스트에 넣을지

너의 추천
호스티드 CI가 그린이면 merge 후보로 둔다. types/config 스플릿 때문에 닫을 PR이 아니고, #5113의 라우티드 남은 구멍을 정확히 닫는 유지보수 PR이다. CI만 확인하고 넣으면 된다.

이 댓글은 grok-bot이 작성했습니다

**Comment by lidge-jun** [comment]
## 추가 리뷰 · 우선순위 78 / 80

이전 리뷰(77/80, tip `641f407c`) 이후 tip이 `bb430bf0`으로 한 커밋 더 갔습니다. 공유 디코더가 라우티드 경로에서 일반 raw 입력을 점진적으로 내보내게 된 뒤, 예전 라우티드 디코더가 “비정규 모양은 전부 홀드”하던 덕분에 안 드러나던 구멍이 열렸습니다. 장식된 `*** Begin Patch ***` 봉투가 미리보기로 나가면, 완료 시점의 `normalizeApplyPatchDelimiters`가 그 마커를 정규화 형태로 바꿔 버립니다. 이미 보낸 장식 바이트를 취소할 SSE가 없으니, 브리지가 이미 하던 것과 같이 라우티드 홀드도 “완료가 봉투를 다시 쓰는 두 이유”를 모두 막아야 합니다.

이번 커밋은 `src/server/responses-custom-tool-repair.ts`에서 `mayCompile`(exec → apply_patch 헬퍼로 재컴파일)과 `mayNormalize`(apply_patch + 소유 grammar → delimiter 정규화)를 나눠 두고, `mayBecomePatchEnvelope(fullInput)`일 때 둘 다 미리보기를 막습니다. 네임스페이스 툴은 예전처럼 그대로입니다. 형제 테스트 `responses-custom-tool-stream-consistency.test.ts`에 raw 장식 봉투를 모든 split 경계로 밀어 넣고, preview는 빈 문자열·done/item/terminal은 정규화된 같은 값·장식 `*** Begin Patch ***`는 최종에 없다고 고정했습니다. `restoreExecStream` 헬퍼는 tool 이름을 받게 넓혀 apply_patch에도 씁니다.

경로 `src/server/responses-custom-tool-repair.ts` (~305-324) - `mayCompile` / `mayNormalize` 분리 후 동일 `mayBecomePatchEnvelope` 게이트. 브리지와 대칭이고, 공유 디코더가 raw를 흘려보낸 뒤 필요한 두 번째 이유를 명시했다
경로 `tests/responses/responses-custom-tool-stream-consistency.test.ts` - 장식 apply_patch 봉투 split 전수 + 정규화 최종값 단언. 회귀 핀으로 충분하다
호스티드 CI - 이 글을 쓰는 시점 hygiene/label/react-doctor/resolve-pr 등이 아직 pending. merge 전에 그린을 기다려야 한다
라인 문제(차단급) - 이 follow-up diff에서 발명할 차단 버그는 없다. 공유 디코더 도입이 연 구멍을 제때 닫은 유지보수다

메인테이너의 판단이 필요한 지점
- 호스티드 CI(특히 responses stream consistency·structure gate)가 그린인지 확인한 뒤 merge할지
- 브리지 `sse.ts`의 패치 봉투 홀드와 이 라우티드 `mayCompile`/`mayNormalize`가 앞으로 다시 어긋나지 않게, “한 디코더 + 대칭 홀드”를 리뷰 체크리스트에 넣을지
- namespaced apply_patch/exec에 같은 정규화가 걸릴 가능성이 생기면 그때 grammar 소유 조건을 다시 볼지(지금 범위에서는 의도적으로 제외)

너의 추천
이전 추천을 유지하되, 이 follow-up까지 포함한 tip(`bb430bf0`) 기준으로 본다. 공유 디코더가 연 raw-progressive 구멍을 브리지와 같은 이유로 막았고 테스트도 있다. 호스티드 CI만 그린이면 merge 후보. types/config 스플릿과 무관하고 #5113을 닫는 유지보수 PR이다.

이 댓글은 grok-bot이 작성했습니다

File: src/server/responses-custom-tool-repair.ts [hunk 1]
```diff
@@ -332,19 +293,26 @@ export function createRoutedCustomToolRestoreBlockRewrite(
...
-      // suppress progressive emission and let the done event carry input.
-      if (FREEFORM_WRAP_PREFIX.startsWith(open.argumentsText)) return [];
-      const fullInput = partialCustomToolInput(open.argumentsText);
+      const itemName = itemNames.get(upstreamItemId);
+      const ownsFreeformGrammar = itemName?.namespace === undefined
+        || itemName?.namespace === "functions";
+      const fullInput = progressiveFreeformInput(
+        open.argumentsText,
+        ownsFreeformGrammar ? itemName?.name ?? "" : "",
+      );
```
[the remark this answers is not in the document]
Comment (lidge-jun) [reply src/server/responses-custom-tool-repair.ts:302]: Fixed at `bb430bf0a1d035261199fe256a32b38630a1aa72` — thank you, this was correct and it was introduced by this PR.

Making ordinary raw input progressive is the intended change, but it exposed a case the older routed decoder never reached, because that decoder held every non-canonical shape. The routed hold now covers both reasons completion rewrites an envelope, mirroring `src/bridge/sse.ts`: `mayCompile` for an `exec` body that becomes an `apply_patch` helper call, and `mayNormalize` for an `apply_patch` body whose delimiters `normalizeApplyPatchDelimiters` rewrites. Namespaced tools are unchanged, since neither reason applies to a foreign grammar.

The regression drives a raw decorated envelope across every chunk split and asserts an empty preview with agreeing normalized done, item and terminal input.

File: src/responses/progressive-freeform-input.ts [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+ *
+ * A lone high surrogate is not a character. Emitting one alone puts an unpaired code unit in a
+ * delta the client has to decode by itself, so the pair is emitted together or not at all.
+ */
+function decodeJsonStringPrefix(body: string): string | null {
+  let out = "";
+  for (let i = 0; i < body.length; i++) {
+    const c = body[i];
+    if (c === '"') break; // unescaped closing quote: value complete
+    if (c !== "\\") { out += c; continue; }
```
[the remark this answers is not in the document]
Comment (lidge-jun) [reply src/responses/progressive-freeform-input.ts]: Fixed at `b1b604e14cfacf7f1efe5f5b6566b89f165a08bc`.

Correct, and it is the same class of defect this PR exists to close: a literal character below U+0020 makes the wrapper unparseable, so completion returns the raw wrapper while the preview had already published a decoded value. The decoder now returns `null` on that branch, with a regression that feeds `{"input":"one<LF>two"}` and asserts an empty preview against raw done, item and terminal input.

File: src/responses/progressive-freeform-input.ts [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+ * The earlier form of this compared the buffer against the compact literal `{"key":"`, so a
+ * wrapper written with spaces or newlines matched no prefix at all, streamed as raw JSON
+ * deltas and then completed as the unwrapped body. That is the same delta/completion
+ * disagreement #5047 closed for compact wrappers, reached through a different spelling:
+ * `unwrapFreeformToolInput` reads the completed text with `JSON.parse`, which does not care
+ * how the object is laid out, so neither can the streaming side.
+ */
+function wrapperOpening(args: string, key: string): WrapperOpening {
+  let index = 0;
+  for (const token of ["{", `"${key}"`, ":", '"']) {
```
[the remark this answers is not in the document]
Comment (lidge-jun) [reply src/responses/progressive-freeform-input.ts:23]: Valid finding, but I do not think this PR is the right place to change it, and I want to be explicit about why rather than quietly skip it.

The gap is real: `wrapperOpening` matches a literal `input` in first position, while `JSON.parse` accepts any order and decodes escaped property names, so `{"metadata":1,"input":"cmd"}` streams raw and completes as `cmd`. That is a preview/completion disagreement of exactly the class this PR closes elsewhere.

It is not introduced here. This decoder is `src/bridge/sse.ts` lifted out verbatim; it shipped in #5070 and this PR's contract is that bridge behavior is byte-identical after the extraction. As you note, the routed caller suppresses unchanged object prefixes, so the reachable path is direct SSE only — the path this PR does not change.

The cheap version of the fix — hold any object-shaped buffer the openings do not recognize — is not available, because the current behavior is deliberately pinned. `tests/adapters/bridge.test.ts` "an ambiguous multi-field object stays unrepaired and byte-exact" asserts that `{"code":"a","script":"b"}` and `{"code":1}` stream raw and complete byte-exact. So closing this properly needs the incremental object scanner you describe, which has to keep both of those true while decoding escaped names and any-order properties. That is its own change with its own regressions, and folding it in here would expand a routed-path fix into a rework of the shared wrapper grammar.

My preference is a follow-up issue against the bridge path, referencing this thread, so the scanner lands with the coverage it needs. Leaving this thread unresolved for the maintainer to decide.

Local checks remain NOT RUN for this lane.

[the remark this answers is not in the document]
Comment (lidge-jun) [reply src/responses/progressive-freeform-input.ts:23]: Tracked separately in #5151. Exact-base review confirms this literal-key behavior predates the extraction; the routed restoration already suppresses unchanged object prefixes, so #5113 acceptance is not weakened. The follow-up specifies bounded property-name scanning, escaped/reordered keys, byte-exact ambiguous non-wrapper objects, chunk partitions and accounting/cancellation coverage. This thread is resolved as separately tracked pre-existing scope; it is not claimed fixed by #5129.

**Comment by lidge-jun** [comment]
Maintainer integration into `dev` under MAINTAINERS.md, without claiming self-approval.

Reviewed head: `b1b604e14cfacf7f1efe5f5b6566b89f165a08bc`. Final independent review covered all nine files and passed. The shared decoder is pure; callers retain budget ownership; raw decorated patch and fenced-wrapper holds, literal-control handling, terminal identity and documented no-rewind limits are covered. The pre-existing direct-bridge reordered/escaped-key limitation is separately tracked in #5151, not represented as fixed here.

Exact-head hosted run [35434780189](https://github.com/lidge-jun/opencodex/actions/runs/35434780189) passed all applicable producer jobs and aggregate `ci`; current control jobs passed. macOS logs show all ten new stream-consistency cases executed and passing. The old cancelled label execution is superseded by a successful current-head label run. Event-inapplicable skips are not claimed as executed tests. Local suites/tests/typecheck/build were NOT RUN.

All review threads are resolved, no maintainer objection remains, and native stack membership is absent. Static merge against current dev `af4f744c75e4dd0348b4e497e67db4ac80208e07` is conflict-free. Integration will be pinned to the reviewed head; #5113 closes only after its acceptance is verified on dev.
